From: Takashi Iwai <tiwai@suse.de>
Subject: Fix the detection of new synaptics touchpad on HP laptops
Patch-mainline: Planned
References: bnc#662212,bnc#663582,bnc#663537,bnc#664463

The new touchpad device gives incompatible capability bits that
screw up the driver.  Adapt the new values temporarily.

Signed-off-by: Takashi Iwai <tiwai@suse.de>

---
 drivers/input/mouse/synaptics.c |    7 +++++--
 1 file changed, 5 insertions(+), 2 deletions(-)

--- a/drivers/input/mouse/synaptics.c
+++ b/drivers/input/mouse/synaptics.c
@@ -138,7 +138,8 @@ static int synaptics_capability(struct p
 	priv->capabilities = (cap[0] << 16) | (cap[1] << 8) | cap[2];
 	priv->ext_cap = priv->ext_cap_0c = 0;
 
-	if (!SYN_CAP_VALID(priv->capabilities))
+	if (!SYN_CAP_VALID(priv->capabilities) &&
+	    (priv->capabilities & 0xffff00) != 0xd00000)
 		return -1;
 
 	/*
@@ -374,7 +375,9 @@ static void synaptics_check_clickpad(str
 	struct synaptics_data *priv = psmouse->private;
 	unsigned char ncap[3];
 
-	if (SYN_CAP_PRODUCT_ID(priv->ext_cap) != 0xe4)
+	if (SYN_CAP_PRODUCT_ID(priv->ext_cap) != 0xe4 &&
+	    (SYN_CAP_PRODUCT_ID(priv->ext_cap) != 0x64 ||
+	     priv->capabilities != 0xd00073))
 		return;
 	if (synaptics_send_cmd(psmouse, 0x0c, ncap))
 		return;
